File: DSA/upama_kumari_590014909/day_55_question_2.py
def longest_unique_substring(s):
    char_index = {}
    left = 0
    max_len = 0

    for right, char in enumerate(s):
        if char in char_index and char_index[char] >= left:
            left = char_index[char] + 1

        char_index[char] = right
        max_len = max(max_len, right - left + 1)

    return max_len

print(longest_unique_substring("abcabcbb")) 
print(longest_unique_substring("bbbbb"))    

File: DSA/Gautam_Shyamnani_590013078/Day55Question2.cpp
#include <iostream>
#include <unordered_set>
#include <string>
using namespace std;

int longestUniqueSubstring(const string& s) {
    unordered_set<char> seen;
    int left = 0, maxLen = 0;

    for (int right = 0; right < s.length(); ++right) {
        while (seen.find(s[right]) != seen.end()) {
            seen.erase(s[left]);
            ++left;
        }
        seen.insert(s[right]);
        maxLen = max(maxLen, right - left + 1);
    }

    return maxLen;
}

int main() {
    string str1 = "abcabcbb";
    cout << "Output: " << longestUniqueSubstring(str1) << endl;

    string str2 = "bbbbb";
    cout << "Output: " << longestUniqueSubstring(str2) << endl;

    return 0;
}
